src/core/control/Control.cpp

```diff
@@ -2979,6 +2986,31 @@ void Control::clipboardPasteXournal(ObjectInputStream& in) {
     }
 }
 
+void Control::moveSelectionToLayer(size_t layerNo) {
+    PageRef currentP = getCurrentPage();
+    if (layerNo < 0 || layerNo >= currentP->getLayerCount()) {
+        return;
+    }
+    auto selection = getWindow()->getXournal()->getSelection();
+    if (!selection) {
+        return;
+    }
+
+    auto selectedElements = selection->getElements();
+    auto oldLayer = currentP->getSelectedLayer();
+    auto newLayer = currentP->getLayers()->at(layerNo);
+    auto moveSelUndo = std::make_unique<MoveSelectionToLayerUndoAction>(currentP, getLayerController(), oldLayer, currentP->getSelectedLayerId() - 1, layerNo);
+    clearSelectionEndText();
+    for (auto e : selectedElements) {
+        currentP->getSelectedLayer()->removeElement(e,false);
+        currentP->getLayers()->at(layerNo)->addElement(e);
+        moveSelUndo->addElement(newLayer, e, newLayer->indexOf(e));
+    }
```

Comment:
So, this:

1. Takes selected elements and add them back onto the original layer (`clearSelectionEndText()`)
2. Removes the elements we just added from the original layer
3. Adds them to the target layer

This add-remove-add sequence is a waste.

Reply:
I agree, this is a weird sequence. Although I'd propose adding the option to not clear the selection when calling `getLayerController()->switchToLay(...)`, since now it basically is a layer switch without clearing the selection (draging the selection with you" while changing layer).
In that case it wouldn't be necessary to change EditSelection and/or EditSelectionContent.

Reply:
> Although I'd propose adding the option to not clear the selection when calling `getLayerController()->switchToLay(...)`

Yes, that sounds good!

Reply:
Then I'll implement it that way and change the UndoAction accordingly

Reply:
In the UndoAction I left the removing and adding to layers as when undoing / redoing the elements might not be selected anymore. It is more intuitive here to just move the elements from the fixed layers.